File: scala/grade-school/src/main/scala/GradeSchool.scala
import scala.collection.mutable.SortedMap

class School {
  var db = SortedMap.empty[Int, Seq[String]]

  def add(name: String, g: Int) = db += (g -> (grade(g) :+ name))

  def grade(g: Int) = db.getOrElse(g, Seq())

  def sorted = db.map(m => m._1 -> m._2.sorted)
}

File: nim/grade-school/grade_school.nim
import sequtils, algorithm
type
  Student = tuple[name: string, grade: int]
  School* = object
    students*: seq[Student]

func `<`(a,b: Student): bool =
  (a.grade, a.name) < (b.grade, b.name)

func roster*(school: School): seq[string] =
    school.students.sorted.mapIt(it.name)

func grade*(school: School, grade: int): seq[string] =
    school.students.filterIt(it.grade == grade)
    .mapIt(it.name).sorted
